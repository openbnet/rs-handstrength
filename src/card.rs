#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub enum Suit {
    S,
    H,
    C,
    D,
    A, // any *
    E, // TrainerSuit first suit in cards
    F, //2nd suit
    G, // 3rd suit
    X, // Except E - 3 possible
    Y, // Except E and F - 2 possible
    Z, // Except E, F, and G -4th suit

}

#[derive(Clone, Copy, Debug, Eq, Hash,PartialOrd, PartialEq, Ord)]
pub struct Card {
    pub value: u8,
    pub suit: Suit,
}
use std::fmt;

impl fmt::Display for Suit {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let suit = match self {
            Suit::C => "c",
            Suit::D => "d",
            Suit::H => "h",
            Suit::S => "s",
            Suit::A => "*",
            Suit::E => "e",
            Suit::F => "f",
            Suit::G => "g",
            Suit::X => "x",
            Suit::Y => "y",
            Suit::Z => "z",
        };
        write!(f, "{}", suit)
    }
}

impl fmt::Display for Card {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} of {}", self.value, self.suit)
    }
}
impl Default for Card {
    fn default() -> Self {
        Card {
            value: 0,
            suit: Suit::S,
        }
    }
}