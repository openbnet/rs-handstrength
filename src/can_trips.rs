use crate::card::{Card, Suit};
use crate::sorting::get_same_value_map;
use itertools::Itertools;
use std::collections::HashSet;
use crate::can_libs::*;

pub fn can_have_trips(comcards: &Vec<Card>) -> CanHaveCombis {
    let mut can_have_combis: CanHaveCombis = Vec::new();
    let grouped_cards = get_same_value_map(comcards);
    let all_values: HashSet<u8> = (2..=14).collect(); // All possible card values
    let has_pair = grouped_cards.iter().any(|group| group.len() == 2);
    let has_trips = grouped_cards.iter().any(|group| group.len() == 3);
    let trips_value = grouped_cards.iter().find(|group| group.len() == 3).map(|group| group[0].value);
    let comcard_values: HashSet<u8> = comcards.iter().map(|card| card.value).collect(); 
    let pair_values: Vec<u8> = grouped_cards.iter()
    .filter(|group| group.len() == 2)
    .map(|group| group[0].value)
    .collect();

    for group in grouped_cards {
        let group_value = group[0].value;

        match group.len() {
            1 if !has_pair && !has_trips => {
                // Single card on board, need to find pairs in the remaining cards
                let possible_pairs = vec![vec![Card { value: group_value, suit: Suit::A }, Card { value: group_value, suit: Suit::A }]];
                can_have_combis.push((group.clone(), vec![possible_pairs]));
            },
            2 => {
                if !has_trips || trips_value.map_or(false, |val| group_value > val) {
                    let mut possible_kickers = all_values.iter()
                        .filter(|&&val| val != group_value && !comcard_values.contains(&val))
                        .map(|&val| Card { value: val, suit: Suit::A })
                        .collect::<Vec<Card>>();
            
                    // Sort kickers in descending order
                    possible_kickers.sort_by(|a, b| b.value.cmp(&a.value));
            
                    let possible_hands = possible_kickers.iter()
                        .map(|kicker| vec![vec![kicker.clone(), Card { value: group_value, suit: Suit::A }]])
                        .collect::<Vec<SameRankHands>>();
            
                    can_have_combis.push((group.clone(), possible_hands));
                }
            },
            
            3 | 4 => {
                // Process trips
                if let Some(t_val) = trips_value {
                    let mut possible_kickers = all_values.iter()
                        .filter(|&&val| val != t_val && (!has_pair || !pair_values.contains(&val)))
                        .map(|&val| Card { value: val, suit: Suit::A })
                        .collect::<Vec<Card>>();
            
                    if has_pair {
                        // Include the lower pair values in the possible kickers
                        for &p_val in &pair_values {
                            if p_val < t_val {
                                possible_kickers.push(Card { value: p_val, suit: Suit::A });
                            }
                        }
                    }
            
                    // Sort kickers in descending order (higher values first)
                    possible_kickers.sort_by(|a, b| b.value.cmp(&a.value));
            
                    // Generate all two-card combinations
                    let possible_hands = possible_kickers.iter()
                        .combinations(2)
                        .map(|combo| vec![combo.iter().map(|&card| *card).collect::<Vec<Card>>()])
                        .collect::<Vec<SameRankHands>>();
            
                    can_have_combis.push((group.clone(), possible_hands));
                }
            },
            
            _ => {} // Ignore groups of size 4 or more
        }
    }

    can_have_combis
}
