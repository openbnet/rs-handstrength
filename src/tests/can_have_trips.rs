#[cfg(test)]
mod tests {
    use crate::card::{Card, Suit};
    use crate::can_trips::can_have_trips;
    use insta::assert_debug_snapshot;
    #[test]
    fn test_can_have_trips_with_trips_on_board() {
        let cards = vec![
            Card { value: 10, suit: Suit::H },
            Card { value: 10, suit: Suit::D },
            Card { value: 10, suit: Suit::C },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
        ];

        let trip_combis = can_have_trips(&cards);
        assert!(!trip_combis.is_empty()); // Ensure we have some combinations
        assert_eq!(trip_combis[0].0.len(), 3); // Check if the board has exactly 3 cards (the trips)
        assert_eq!(trip_combis[0].0[0].value, 10); // Check if the board cards are the trips (value 10)
        // Add more assertions as needed
        assert_debug_snapshot!(trip_combis);
    }

    #[test]
    fn test_can_have_trips_without_trips_on_board() {
        let cards = vec![
            Card { value: 10, suit: Suit::H },
            Card { value: 9, suit: Suit::D },
            Card { value: 8, suit: Suit::C },
            Card { value: 2, suit: Suit::S },
            Card { value: 3, suit: Suit::S },
        ];

        let trip_combis = can_have_trips(&cards);
        // assert!(trip_combis.is_empty()); // Expecting no combinations
        // println!("no pair trips {:?}", trip_combis);
        assert_debug_snapshot!(trip_combis);
    }

    #[test]
    fn test_can_have_trips_with_various_card_combinations() {
        let cards = vec![
            Card { value: 2, suit: Suit::H },
            Card { value: 2, suit: Suit::D },
            Card { value: 2, suit: Suit::C },
            Card { value: 5, suit: Suit::S },
            Card { value: 7, suit: Suit::S },
        ];

        let trip_combis = can_have_trips(&cards);
        assert!(!trip_combis.is_empty()); // Expecting some combinations
        // Add specific checks for expected combinations
        assert_debug_snapshot!(trip_combis);
    }
    #[test]
    fn test_can_have_trips_with_fh() {
        let cards = vec![
            Card { value: 2, suit: Suit::H },
            Card { value: 2, suit: Suit::D },
            Card { value: 2, suit: Suit::C },
            Card { value: 5, suit: Suit::S },
            Card { value: 5, suit: Suit::S },
        ];

        let trip_combis = can_have_trips(&cards);
        assert!(!trip_combis.is_empty()); // Expecting some combinations
        // Add specific checks for expected combinations
        assert_debug_snapshot!(trip_combis);
    }
    #[test]
    fn test_can_have_trips_with_over_fh() {
        let cards = vec![
            Card { value: 2, suit: Suit::H },
            Card { value: 2, suit: Suit::D },
            Card { value: 5, suit: Suit::C },
            Card { value: 5, suit: Suit::S },
            Card { value: 5, suit: Suit::S },
        ];

        let trip_combis = can_have_trips(&cards);
        assert!(!trip_combis.is_empty()); // Expecting some combinations
        // Add specific checks for expected combinations
        assert_debug_snapshot!(trip_combis);
    }
}
