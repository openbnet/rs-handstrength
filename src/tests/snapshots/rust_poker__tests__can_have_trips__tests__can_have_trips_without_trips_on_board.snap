---
source: src/tests/can_have_trips.rs
expression: trip_combis
---
[
    (
        [
            Card {
                value: 10,
                suit: H,
            },
        ],
        [
            [
                [
                    Card {
                        value: 10,
                        suit: A,
                    },
                    Card {
                        value: 10,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 9,
                suit: D,
            },
        ],
        [
            [
                [
                    Card {
                        value: 9,
                        suit: A,
                    },
                    Card {
                        value: 9,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 8,
                suit: C,
            },
        ],
        [
            [
                [
                    Card {
                        value: 8,
                        suit: A,
                    },
                    Card {
                        value: 8,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 3,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 3,
                        suit: A,
                    },
                    Card {
                        value: 3,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
    (
        [
            Card {
                value: 2,
                suit: S,
            },
        ],
        [
            [
                [
                    Card {
                        value: 2,
                        suit: A,
                    },
                    Card {
                        value: 2,
                        suit: A,
                    },
                ],
            ],
        ],
    ),
]
